week2/LogAnalysis.hs

```diff
@@ -0,0 +1,56 @@
+{-# OPTIONS_GHC -Wall #-}
+
+module LogAnalysis where
+
+import Log
+
+-- EXERCISE 1 --
+
+parseMessage :: String -> LogMessage
+parseMessage = 
+  convert . words
+  where convert :: [String] -> LogMessage
+        convert ("I":t:m) = LogMessage Info (read t) (unwords m)
+        convert ("W":t:m) = LogMessage Warning (read t) (unwords m)
+        convert ("E":c:t:m) = LogMessage (Error (read c)) (read t) (unwords m)
+        convert x = Unknown (unwords x)
+
+parse :: String -> [LogMessage]
+parse = map parseMessage . lines
+
+-- EXERCISE 2 --
+
+insert :: LogMessage -> MessageTree -> MessageTree
+insert (Unknown _) tree = tree
+insert newLog Leaf = Node Leaf newLog Leaf
+insert newLog (Node left root right) = 
+  let t1 = getTime newLog
+      t2 = getTime root
+  in if t1 <= t2
+    then Node (insert newLog left) root right
+    else Node left root (insert newLog right)
```

Comment:
I'd avoid using `if` statements for simple statements, you can use one of these syntaxes instead:
```
insert newLog@(LogMessage _ t1 _ ) (Node left root@(LogMessage _ t2 _ ) right)
  | t1 <= t2 = ...
  | otherwise = ...
```

```
case compare t1 t2 of
    LT -> ...
    _   -> ...
```
(compare returns something of type [Ordering](https://hackage.haskell.org/package/base-4.12.0.0/docs/Prelude.html#t:Ordering): `LT, EQ, GT`)